### DISCUSSION TITLE
Offset based paging in GraphQL integration

### DISCUSSION THREAD

**Discussion by david-kubecka** [Blazebit/blaze-persistence#1562] [Q&A] [answered]
I would like to migrate our existing app from a Spring Data + GraphQL-specific-DTOs setup to Blaze persistence, utilizing EntityViews instead of the DTOs. Using the documentation (btw it's great!) + example projects I was able to figure out most of the stuff with the exception of paging.

Currently, we use the classic offset-based paging and pass `Pageable` into the repository `findAll` methods. While I would like to utilize the provided Keyset pagination in the future, this is a breaking change of the API and as such, I would like to postpone that until later. The question therefore is: how I can use the offset paging mechanism with the GraphQL integration?

From the docs it seems that all paging input data are encapsulated in the `DataFetchingEnvironment`, which however comes from the client request and doesn't seem to be modifiable. Is there any other way how to simply pass offset/limit parameters to the entity view setting? E.g. similarly to `EntityViewSetting.create(CatView.class, 0, 10)` but with the ability to pass `DataFetchingEnvironment`?

**Comment by david-kubecka** [comment, marked as answer]
Inspecting the source code I think I see the solution now. There is this public method in `GraphQLEntityViewSupport`:
```
public <T> EntityViewSetting<T, CriteriaBuilder<T>> createSetting(Class<T> entityViewClass, DataFetchingEnvironment dataFetchingEnvironment, String elementRoot, KeysetPage keysetPage, Integer first, Integer last, Integer offset)
```
where I can simply pass `first` and `offset` (other paging related params will be null).

**Reply by beikov** [reply]
Actually, even the `createPaginatedSetting(DataFetchingEnvironment dataFetchingEnvironment)` will work. If you read the documentation, you will see that it expects fields with certain names in the GraphQL query, from which it will read the pagination values.

**Reply by david-kubecka** [reply]
Yeah, I've seen that. But that will require an API change which - as I said - I want to avoid for now. Correct me if I'm wrong.

**Reply by beikov** [reply]
You mean that you have chosen different names for your offset/limit parameters? If so, the only sensible way for you to model this would be by extending `GraphQLEntityViewSupportFactory` and construct the `GraphQLEntityViewSupport` with the custom names. I see that it's a bit hard to do that right now. I can extract a method for constructing the `GraphQLEntityViewSupport` so that you only have to override that one. Would that help in your case?

**Reply by david-kubecka** [reply]
Actually, we wrap offset/limit into a specific type in the input so you would rather have to provide a more general `PagingInputExtractor` interface which will let us extract the wrapper paging type into scalars. But that seems to be quite an overkill for me, so if you don't see any problem with my suggested option I will go in that direction and try to switch to the supported keyset paging mechanism ASAP.

**Reply by beikov** [reply]
Yeah that's an overkill. I tried to rely on what the Relay specification requires, as this seemed most natural to me. Thanks for the clarification.
